contracts/Staking.sol

```diff
@@ -0,0 +1,466 @@
+pragma solidity 0.6.4;
+
+import "./System.sol";
+import "./lib/Memory.sol";
+import "./lib/BytesToTypes.sol";
+import "./interface/IApplication.sol";
+import "./interface/ICrossChain.sol";
+import "./interface/IParamSubscriber.sol";
+import "./interface/ITokenHub.sol";
+import "./lib/CmnPkg.sol";
+import "./lib/SafeMath.sol";
+import "./lib/RLPEncode.sol";
+import "./lib/RLPDecode.sol";
+import "./interface/IStaking.sol";
+
+contract Staking is IStaking, System, IParamSubscriber, IApplication {
+  using SafeMath for uint256;
+  using RLPEncode for *;
+  using RLPDecode for *;
+
+  struct DelegateAckPackage {
+    address delegator;
+    address validator;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct UndelegateAckPackage {
+    address delegator;
+    address validator;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct RedelegateAckPackage {
+    address delegator;
+    address valSrc;
+    address valDst;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct DistributeRewardSynPackage {
+    uint256 amount;
+    address recipient;
+  }
+
+  struct DistributeUndelegatedSynPackage {
+    uint256 amount;
+    address recipient;
+    address validator;
+  }
+
+  // Cross-Chain Stake Event type
+  uint8 public constant EVENT_DELEGATE = 0x01;
+  uint8 public constant EVENT_UNDELEGATE = 0x02;
+  uint8 public constant EVENT_REDELEGATE = 0x03;
+  uint8 public constant EVENT_DISTRIBUTE_REWARD = 0x04;
+  uint8 public constant EVENT_DISTRIBUTE_UNDELEGATED = 0x05;
+
+  uint32 public constant ERROR_UNKNOWN_PACKAGE_TYPE = 101;
+  uint32 public constant ERROR_WITHDRAW_BNB = 102;
+
+  uint256 constant public TEN_DECIMALS = 1e10;
+
+  uint256 public constant INIT_ORACLE_RELAYER_FEE = 2e16; //TODO
+  uint256 public constant INIT_MIN_DELEGATION_CHANGE = 1e20;
+
+  uint256 public oracleRelayerFee;
+  uint256 public minDelegationChange;
```

Comment:
I know in Beacon Chain, it is called `minDelegationChange`, but I think `minDelegation` is good for understanding.

Reply:
OK